File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
- Use a hash Map data structure to retain the copy of the link list
- Create a copy of each pointer of the linked list and store it in a hash map
- Build the next pointer and the random pointer using the hash Map
"""

"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        oldToCopy = {None:None}
        
        # Build a copy of linked List
        cur = head
        while cur:
            copy = Node(cur.val)
            oldToCopy[cur] = copy
            cur = cur.next
            
        cur = head
        while cur:
            copy = oldToCopy[cur]
            copy.next = oldToCopy[cur.next]
            copy.random = oldToCopy[cur.random]
            cur = cur.next
            
        return oldToCopy[head]

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    copyToOld := make(map[*Node]*Node)
    
    cur := head
    for cur != nil {
        copy := &Node{Val: cur.Val}
        copyToOld[cur] = copy
        cur = cur.Next
    }
    cur = head
    for cur != nil {
        copy := copyToOld[cur]
        copy.Next = copyToOld[cur.Next]
        copy.Random = copyToOld[cur.Random]
        cur = cur.Next
    }
    
    return copyToOld[head]
}